dds/DCPS/XTypes/DynamicData.cpp

```diff
@@ -0,0 +1,2169 @@
+/*
+ * Distributed under the OpenDDS License.
+ * See: http://www.opendds.org/license.html
+ */
+
+#include <DCPS/DdsDcps_pch.h>
+
+#include "DynamicData.h"
+
+#include "DynamicTypeMember.h"
+
+#include <ace/OS_NS_string.h>
+
+#include <stdexcept>
+
+OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
+
+namespace OpenDDS {
+namespace XTypes {
+
+DynamicData::DynamicData()
+  : chain_(0)
+  , encoding_(DCPS::Encoding::KIND_XCDR2)
+  , reset_align_state_(false)
+  , strm_(0, encoding_)
+  , item_count_(ITEM_COUNT_INVALID)
+{}
+
+DynamicData::DynamicData(ACE_Message_Block* chain,
+                         const DCPS::Encoding& encoding,
+                         const DynamicType_rch& type)
+  : chain_(chain->duplicate())
+  , encoding_(encoding)
+  , reset_align_state_(false)
+  , strm_(chain_, encoding_)
+  , type_(get_base_type(type))
+  , descriptor_(type_->get_descriptor())
+  , item_count_(ITEM_COUNT_INVALID)
+{
+  if (encoding_.xcdr_version() != DCPS::Encoding::XCDR_VERSION_2) {
+    throw std::runtime_error("DynamicData only supports XCDR2 at this time");
+  }
+}
+
+DynamicData::DynamicData(DCPS::Serializer& ser, const DynamicType_rch& type)
+  : chain_(ser.current()->duplicate())
+  , encoding_(ser.encoding())
+  , reset_align_state_(true)
+  , align_state_(ser.rdstate())
+  , strm_(chain_, encoding_)
+  , type_(get_base_type(type))
+  , descriptor_(type_->get_descriptor())
+  , item_count_(ITEM_COUNT_INVALID)
+{
+  if (encoding_.xcdr_version() != DCPS::Encoding::XCDR_VERSION_2) {
+    throw std::runtime_error("DynamicData only supports XCDR2 at this time");
+  }
+}
+
+DynamicData::DynamicData(const DynamicData& other)
+  : strm_(0, other.encoding_)
+{
+  copy(other);
+}
+
+DynamicData& DynamicData::operator=(const DynamicData& other)
+{
+  if (this != &other) {
+    copy(other);
+  }
+  return *this;
+}
+
+DynamicData::~DynamicData()
+{
+  ACE_Message_Block::release(chain_);
+}
+
+void DynamicData::copy(const DynamicData& other)
+{
+  chain_ = other.chain_->duplicate();
```

Comment:
Since `chain_` may be non-null when entering this method, this needs an `ACE_Message_Block::release()` first.